File: Baekjoon/yerim/톱니바퀴.cpp
// [14891] 톱니바퀴
#include <iostream>
#include <vector>
using namespace std;

int k;
vector <string> state, tmp;

void rotateOneWheel(int idx, int dir) 
{
    if(dir == 1) {
        int num = tmp[idx][7];
        tmp[idx].pop_back();
        tmp[idx].insert(tmp[idx].begin(), num);
    }
    else {
        int num = tmp[idx][0];
        tmp[idx].erase(tmp[idx].begin());
        tmp[idx].push_back(num);
    }
}

bool leftSide(int compare, int idx, int dir)
{
    if(state[compare-1][6] != state[idx-1][2]) {
        rotateOneWheel(idx-1, dir);
        return true;
    }
    return false;
}

bool rightSide(int compare, int idx, int dir)
{
    if(state[compare-1][2] != state[idx-1][6]) {
        rotateOneWheel(idx-1, dir);
        return true;
    }
    return false;
}

void go(int num, int dir)
{
    switch (num) {
        case 1:
            if(rightSide(1, 2, -1*dir)) {
                if(rightSide(2, 3, dir)) {
                    rightSide(3, 4, -1*dir);
                }
            }
            break;
        case 2:
            leftSide(2, 1, -1*dir);
            if(rightSide(2, 3, -1*dir)) {
                rightSide(3, 4, dir);
            }
            break;
        case 3:
            if(leftSide(3, 2, -1*dir)) {
                leftSide(2, 1, dir);
            }
            rightSide(3, 4, -1*dir);
            break;
        case 4:
            if(leftSide(4, 3, -1*dir)) {
                if(leftSide(3, 2, dir)) {
                    leftSide(2, 1, -1*dir);
                }
            }
        default: break;
    }
    rotateOneWheel(num-1, dir);
}

void input()
{
    int num, dir;
    string s;
    for(int i = 0; i < 4; i++) {
        cin >> s;
        state.push_back(s);
        tmp.push_back(s);
    }
    cin >> k;
    for(int i = 0; i < k; i++) {
        cin >> num >> dir;
        go(num, dir);
        copy(tmp.begin(), tmp.end(), state.begin());
    }
}

int getScore()
{
    int ans = 0;
    if(state[0][0] == '1') ans += 1;
    if(state[1][0] == '1') ans += 2;
    if(state[2][0] == '1') ans += 4;
    if(state[3][0] == '1') ans += 8;

    return ans;
}

int main(void)
{
    input();
    cout << getScore() << '\n';

    return 0;
}

File: Baekjoon/dongwan/톱니바퀴.py
t1 = list(map(int,input()))
t2 = list(map(int,input()))
t3 = list(map(int,input()))
t4 = list(map(int,input()))

k = int(input())

#맞닿은 부분은
#1번-> 2 
#2번 -> 2,-2 
#3번 -> 2,-2 
#4번 -> -2

def get_rotate(target) :
    if target == 1 :
        if t1[2] != t2[-2] :
            tmp.append([2,t2])
    elif target == 2 :
        if t1[2] != t2[-2] :
            tmp.append([1,t1])
        if t2[2] != t3[-2] :
            tmp.append([3,t3])
    elif target == 3 :
        if t2[2] != t3[-2] :
            tmp.append([2,t2])
        if t3[2] != t4[-2] :
            tmp.append([4,t4])
    else :
        if t3[2] != t4[-2] :
            tmp.append([3,t3])
    
print(t1,t2,t3,t4)
for rotation in range(k) :
    target,dir = map(int,input().split())
    
    tmp = []
    get_rotate(target) 
    for n,t in tmp :
        if dir == 1 :
            t = [t[-1]]+t[:7]
        else :
            t = t[1:] + [t[-1]]
        
        if n == 1 :
            t1 = t
        elif n == 2 :
            t2 = t
        elif n == 3 :
            t3 = t 
        else :
            t4 = t 
print(t1,t2,t3,t4)
ans = 0
if t1[0] :
    ans += 1 
if t2[0] :
    ans += 2 
if t3[0] :
    ans += 4 
if t4[0] :
    ans += 8
print(ans)
        